_template_python/INDICATOR_DEV_GUIDE.md

```diff
@@ -0,0 +1,446 @@
+
+If you want to map from one of these to another, the [`delphi_utils.geomapper`](https://github.com/cmu-delphi/covidcast-indicators/blob/6912077acba97e835aff7d0cd3d64309a1a9241d/_delphi_utils_python/delphi_utils/geomap.py) utility covers most cases.
+A brief example of adding states with their population:
+
+```{python}
+from delphi_utils.geomap import GeoMapper
+geo_mapper = GeoMapper()
+geo_mapper.add_geocode(df, "state_id", "state_code", from_col = "state") # add codes and ids from the full names
+df = geo_mapper.add_population_column(df, "state_code") # add state populations
+hhs_version = geo_mapper.replace_geocode(df, "state_code","hhs", new_col = "geo_id") # aggregate to hhs regions, renaming the geo column to geo_id
+```
+
+This example is taken from [`hhs_hosp`](https://github.com/cmu-delphi/covidcast-indicators/blob/main/hhs_hosp/delphi_hhs/run.py); more documentation can be found in the `geomapper` class definition.
+
+#### Implement a Missing Value code system
+
+The column is described [here](https://cmu-delphi.github.io/delphi-epidata/api/missing_codes.html).
+
+#### Testing
+
+As a general rule, it helps to decompose your functions into operations for which you can write unit tests.
+To run the tests, use `make test` in the top-level indicator directory.
+
+Unit tests are required for all functions.
+Integration tests are highly desired, but may be difficult to set up depending on where the data is being fetched from.
+Mocking functions are useful in this case.
+
+#### Naming
+
+Indicator and signal names need to be approved by [@RoniRos](https://www.github.com/RoniRos).
+
+The data source name as specified during an API call (e.g. in `epidatr::pub_covidcast(source = "jhu-csse", ...)`, "jhu-csse" is the data source name) should match the wildcard portion of the module name ("jhu" in `delphi_jhu`) _and_ the top-level directory name in `covidcast-indicators` ("jhu").
+(Ideally, these would all also match how we casually refer to the indicator ("JHU"), but that's hard to foresee and enforce.)
+
+Ideally, the indicator name should:
+
+* Make it easy to tell where the data is coming from
+* Make it easy to tell what type of data it is and/or what is unique about it
+* Be uniquely identifying enough that if we added another indicator from the same organization, we could tell the two apart
+* Be fairly short
+* Be descriptive
+
+Based on these guidelines, the `jhu-csse` indicator would be better as `jhu-csse` everywhere (module name could be `delphi_jhu_csse`), rather than having a mix of `jhu-csse` and `jhu`.
+
+Signal names should not be too long, but the most important feature is that they are descriptive.
+
+Some standard tags used in signal names:
+
+* `raw`: unsmoothed, _no longer used; if no smoothing is specified the signal is assumed to be "raw"_
+* `7dav`: smoothed using a average over a rolling 7-day window; comes at the end of the name
+* `smoothed`: smoothed using a more complex smoothing algorithm
```

Comment:
I'm trying to introduce these "tags" to describe how we've historically named signals, and also how we want to name signals in the future.

Example historical names:

JHU raw: confirmed_cumulative_prop
JHU smoothed: confirmed_7dav_cumulative_prop

Google raw: s01_raw_search
Google smoothed: s01_smoothed_search

Change smoothed: smoothed_adj_outpatient_covid

HHS raw: confirmed_admissions_covid_1d
HHS smoothed: confirmed_admissions_covid_1d_7dav

So "smoothed" shows up in various places. "7dav" has a similar meaning and in my mind should be in the same position. It also shows up in various positions in signal names but most consistently shows up at the end, as in the HHS smoothed signal.

Does that mean that we should say "smoothed" also comes at the end of the name, so that people use it like that in the future? I don't think any existing names put it at the end.

I haven't come up with an entire naming strategy, just some general tips. Do you have thoughts on naming format? Or should we leave it kind of open-ended?

Reply:
oh, I guess I just didn't do a thorough enough survey to notice it showed up in the middle sometimes.

I would change the wording to be prescriptive if that's what you're trying to communicate, and saying it should be at the end does sound like a good idea.

It may be worth mentioning if we're mirroring another processed dataset, that keeping their naming conventions is probably preferable.